workspaces/adventure-pack/goodies/typescript/Map.groupBy/index.test.ts

```diff
@@ -0,0 +1,157 @@
+  });
+
+  it("should work with custom iterables", () => {
+    const customIterable = {
+      *[Symbol.iterator]() {
+        yield 1;
+        yield 2;
+        yield 3;
+      },
+    };
+    const result = Map.groupBy(customIterable, (num) =>
+      num % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result.size).toBe(2);
+    expect(result.get("even")).toStrictEqual([2]);
+    expect(result.get("odd")).toStrictEqual([1, 3]);
+  });
+
+  it("should work with generators", () => {
+    function* numberGenerator() {
+      yield 1;
+      yield 2;
+      yield 3;
+    }
+    const result = Map.groupBy(numberGenerator(), (num) =>
+      num % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result.size).toBe(2);
+    expect(result.get("even")).toStrictEqual([2]);
+    expect(result.get("odd")).toStrictEqual([1, 3]);
+  });
+
+  it("should work with Maps", () => {
+    const map = new Map([
+      ["a", 1],
+      ["b", 2],
+      ["c", 3],
+    ]);
+    const result = Map.groupBy(map, ([_, value]) =>
+      value % 2 === 0 ? "even" : "odd",
+    );
+
+    expect(result.size).toBe(2);
+    expect(result.get("even")).toStrictEqual([["b", 2]]);
+    expect(result.get("odd")).toStrictEqual([
+      ["a", 1],
+      ["c", 3],
+    ]);
+  });
+
+  it("should handle objects with custom iterator", () => {
+    const obj: Iterable<number> = {
+      *[Symbol.iterator]() {
+        const data = [1, 2, 3, 4];
+        for (const item of data) {
+          yield item;
+        }
+      },
+    };
+    const result = Map.groupBy(obj, (num) => (num % 2 === 0 ? "even" : "odd"));
+
+    expect(result.size).toBe(2);
+    expect(result.get("even")).toStrictEqual([2, 4]);
+    expect(result.get("odd")).toStrictEqual([1, 3]);
+  });
+
+  it("should throw TypeError for non-iterable arguments", () => {
+    expect(() => {
+      // @ts-ignore - Intentionally passing invalid type
+      Map.groupBy(123, (x) => x);
+    }).toThrow(TypeError);
+  });
+
+  it("should handle undefined or null keys", () => {
+    const data = [1, 2, 3, 4, 5];
+    const result = Map.groupBy(data, (num) =>
+      num % 2 === 0 ? undefined : null,
+    );
+
+    expect(result.size).toBe(2);
+    expect(result.get(undefined)).toStrictEqual([2, 4]);
+    expect(result.get(null)).toStrictEqual([1, 3, 5]);
+  });
+
+  it("should handle keys that are objects or arrays", () => {
+    const data = [1, 2, 3, 4];
+    const even = { even: true };
+    const odd = ["odd"];
+    const result = Map.groupBy(data, (num) => (num % 2 === 0 ? even : odd));
+
+    expect(result.size).toBe(2);
+    expect(result.get({ even: true })).toBeUndefined();
+    expect(result.get(["odd"])).toBeUndefined();
+    const keys = [...result.keys()];
+    expect(
+      keys.some(
+        (key) => typeof key === "object" && "even" in key && key.even === true,
+      ),
+    ).toBeTruthy();
+    expect(
+      keys.some((key) => Array.isArray(key) && key[0] === "odd"),
+    ).toBeTruthy();
```

Comment:
This one might be worth a full-fledged `toStrictEqual` as well.

Reply:
Totally, good call.